README.md
Network Unix from University of Illinois.

ARNAPET NCP additions for Unix V6.


tape1-files/history
unix.32v	30 Nov 79	agn	Original Bell UNIX

u1.1.lp		30 Nov 79	agn	Moved swap to drive 1, added 16 line
					dz, added lp device.  Changed dz.c,
					conf.c, univec.c. Also, changed tty 
					mode NL3 to provide a pad of 020 after
					a new line for Ann Arbor.  Changed
					tty.c. Updated makefile and loadall.
					Moved dev/lp.c to sys/lp.c.

u1.2.lp		07 Dec 79	fram	Removed conversion of newline to
					formfeed from line printer driver
					(sys/lp.c).  Was causing whole
					blank pages on line printer listings.

u1.3.dz         12 Dec 79       fram    Commented out all references to modem
				        control bits in dz.c.  This was a  
					quick-and-dirty fix to make dz driver
					work with current-loop lines, for which
					modem status is undefined.  Does not
					hurt in EIA case either.  (Essentially,
					all our lines are "hard-wired", so 
					modem control is really unnecessary.)

u1.4.lp         18 Dec 79       fram    Edited lp.c to have default line 
					printer width of 80 columns (had been
					changed earlier to 132).  Note:  
					observed behavior after change is that
					line printer prints a max of 76 cols.
					This is a mystery.

u1.5.tty         4 Jan 80       edh     Changed tty.c to throw away extra
					X-ON X-OFF characters when reading
					from a terminal.

u1.6.mknod      11 Jan 80       fram    Changed mknod system call (sys2.c)
					so that third argument is not coerced
					to a short.  Allows specification of
					a generalized first block pointer
					which is useful for creating an inode
					which "captures" a specific bad block.

u1.7.tty        24 Jan 80       fram    Changed tty.c to handle X-ON/X-OFF  
					properly at interrupt level.  
					Supersedes fix made by edh in u1.5.tty
					on 4 Jan 80.

u1.8.tty        9 Feb 80        dbarach, Changed tty.c to cause anything
				rwells  on the tty input queue to cause
					gtty to set the 0x10000 bit in the
					flag word. This bit was unused by 
					the previous system. This is a quick
					and dirty empty() call for use in
					the editor.

-------------------- sys.2 created  9 May 80 -------------------------------


2       15 May 80 tpb and yba   u2.1.mem 
	change mem.c driver to test against PHYSPAGES, not MAXMEM
	Change to h/file.h: add flag values for ncp.
	Add ncpk subdirectory.
	Add ncp header files to h subdirectory.

2       9 May 80 u2.2.ncp       yba     add ncp to v7 vax unix
	Change to h/file.h: add flag values for ncp.
	Add ncpk subdirectory.
	Add ncp header files to h subdirectory.
	Change use of spl, mapalloc, buf headers in ncp code.  No RMI.
	Change file structures (h/file.h).
	Add define of spl_imp as spl5 to h/netparam.h
	Add iialloc/free routines to sys/alloc.c
	Add fushort sushort routines to sys/locore.s
	Add entry for acc_iint/acc_oint in sys/univec.c

2       1 Aug 80 u2.3.ncp       yba     add hooks to nrdwr
	Change sys/sys2.c for network.

2       7 Aug 80 u2.4.ncp       yba     fix bug in kerbuf
	Change to kerbuf.c: use geteblk() in place of old getblk(NODEV,0).

________________________________ Converted to Berkeley System  10-Aug-80 ___

------------------------------ Sys.3 created 12 August 1980 ---------------

3       12 Aug 80 vmu3.1.ncp      yba     add ncp kernel to UC/B system
	Add header files to h subdirectory:
		acc.h
		contab.h
		hosthost.h
		ill.h
		imp.h
		imp11a.h
		iobuf.h
		leader.h
		ncp.h
		net.h
		netbuf.h
		netopen.h
		netparam.h
		rawnet.h
		rfnm.h
		leader.h

	changes to h/
		param.h
		buf.h
		user.h
		file.h

	Add directories ncpk, ncpd, smalldaemon

	changes to ncpk/
		acc.c

	Changes to sys/
		alloc.c         add iialloc/iifree routines
		locore.s        add fushort, sushort
		univec.c        add acc interrupt service
		sys2.c
		conf.c          (Change major device number on /dev/ncpkernel)

3       vmu3.2.ncp      yba     13 Aug 80       fix buffer sizes to vmunix size
	Changes to h/netparam.h (kb_hiwat and init kb_), h/netbuf.h

3       vmu3.2.test     yba     24 Sep 90       disable imp hardware for test

3       vmu3.3.test     agn     21 Oct 80       added modifications to 
                                                sys/fio.c, sys/sys2.c, 
                                                sys/sys3.c to invoke network.
						Adopted C/70 network code for 
						kernel.

3       vmu3.4.hog      dan     ?               ?

3       vmu3.5.tty      rdk     12 Dec 80       changed padding on -nl2
						from 6 to 2 for jp

--------------------------------------- sys.4 created 5 January 81 -------------

4       vmu4.1.test     rfg     5 Jan 81        Added TCP network code
						to kernel.
						mods to sys/sys2.c, sys/sys3.c,
						sys/fio.c, sys/conf.c
						added h/acc.h h/fsm.h h/fsmdef.h
						h/ip.h h/net.h h/tcp.h
						h/tcp_pred.h and net/*.c

4       vmu4.2.loop     rfg     12 Jan 81       added s/w looping to imp_io.c
						test version

4       vmu4.3.tcp      rfg     2 March 81      first production tcp kernel.

4       vmu4.7.pty      rfg     23 March 81     added ptys (pty.c)

----------------------------sys.5 created May 20 1981--------------------------
5	vmu5.0.tty	ers	11 Aug 81	Change tty driver so that it
						does not echo the extra break
						character (t_brkc) if that
						character is ESC (octal 33).
						Change to ptcwrite in pty.c
						to call
						linesw[tp->t_line].l_rint
						instead of ttyinput.


nosc-files/h/tty.h
#
/* Revision - New handshake handler */

/*
 * A clist structure is the head
 * of a linked list queue of characters.
 * The characters are stored in 4-word
 * blocks containing a link and 6 characters.
 * The routines getc and putc (m45.s or m40.s)
 * manipulate these structures.
 */
struct clist
{
	int	c_cc;		/* character count */
	int	c_cf;		/* pointer to first block */
	int	c_cl;		/* pointer to last block */
};

/*
 * A tty structure is needed for
 * each UNIX character device that
 * is used for normal terminal IO.
 * The routines in tty.c handle the
 * common code associated with
 * these structures.
 * The definition and device dependent
 * code is in each driver. (kl.c dc.c dh.c)
 */
struct tty
{
	/* These variables are (hopefully) untouchable by the ttymode system
	 * call. The total number of un-hackable bytes should be specified in
	 * the compiler control line for FIXOFS, below.
	 */

	struct	clist t_canq;	/* buffer for ttyread */
	struct	clist t_outq;	/* output list to device */

	int	*t_addr;		/* device address (register or startup fcn) */
	int 	t_dev;			/* device name */
	int	t_llf;			/* ptr to last linefeed in t_canq */
	int	t_state;		/* internal state, not visible externally */
	char	t_char;			/* character temporary */
	char	t_acnt;			/* count of characters sent */
	int	t_pgrp;			/* process group name */
#ifdef	DNTTY
	int	t_dnflgs;	/* DN: TTY flags specific to DN hacks */
	int	t_ioct;		/* DN: I/O cnt for catching idle lines */
#endif

#define TFIXOFS 28		/* Fixed offset from base of tty structure of
				 * first user-hackable parameter, for ttymod */

	/* User hackable info. Any additional parameters to the rear, please. */

	/*	name		       byte #	      Description */

	int	t_speeds;	/*	0,1	input (low) + output (high) speeds */
	char	t_erase;	/*	2	erase char */
	char	t_kill;		/*	3	kill char */
	int	t_flags;	/*	4,5	mode. Bytes 0-5 settable by stty call */
	int 	t_flag2;	/*	6,7	additional flags */
	char	t_flag4;	/*	8	additional bits */
	char	t_cano;		/*	9	canonical terminal type */
	char	t_clin;		/*	10	current line of device */
	char	t_col;		/*	11	printing column of device */
	char 	t_llen;		/*	12	line length */
	char	t_psize;	/*	13	page size */
	char	t_quit;		/*	14	quit char (^B) */
	char	t_intr;		/*	15	interrupt char (^Q) */
	char	t_EOT;		/*	16	end_of_file char (^D) */
	char	t_cntla;	/*	17	special interrupt char (^A) */
	char	t_atime;	/*	18	timeout in 60ths of seconds */
	char	t_asize;	/*	19	number of character per ACK */
	char	t_enq;		/*	20	ENQ character */
	char	t_ack;		/*	21	ACK character */
	char	t_esc;		/*	22	Escape character */
	char	t_alwat;	/*	23	Low water mark for ACK */
	char	t_delpos;	/*	24	Position of first typed char */
	char	t_freeze;	/*	25	freeze character (^F) */
	char	t_thaw;		/*	26	thaw character (^T) */
	char	t_hold;		/*	27	hold character (^S) */
	char	t_release;	/*	28	release character (^Q) */
	char	t_retype;	/*	29	retype line (^R) */
	char	t_silent;	/*	30	silence output (^S) */
	char	t_clear;	/*	31	clear screen (^L) */

#define TMAXBN	31		/* Maximum allowable byte number for ttymod */
};

char partab[];			/* ASCII table: parity, character class */

#define	TTIPRI	10
#define	TTOPRI	20
#define	CERASE	0177		/* default special characters */
#define	CEOT	032		/* ^Z */
#define	CKILL	('U' & 037)
#define	CQUIT	('B' & 037)		/* FS, cntl shift L */
#define	CINTR	('C' & 037)		/* ^C */
#define CSPEC	('A' & 037)		/* Special interrupt, ^A */
#define CACK	('F' & 037)
#define CENQ	('D' & 037)
#define CESC	('_' & 037)
#define CFREEZE	('S' & 037)
#define CTHAW	('Q' & 037)
#define CHOLD	('S' & 037)
#define CREL	('Q' & 037)
#define CRETYPE	('R' & 037)
#define CCLEAR	('L' & 037)
#define CSILENT	('O' & 037)
#define CESC1	'1'
#define CESC2	'2'

/* limits */
#define	TTHIWAT	50
#define	TTLOWAT	30
#define	TTYHOG	256
#define TTACK	50

/* modes */
#define	HUPCL	01
#define	XTABS	02
#define	LCASE	04
#define	ECHO	010
#define	CRMOD	020
#define	RAW	040
#define	ODDP	0100
#define	EVENP	0200
#define	NLDELAY	001400
#define	TBDELAY	006000
#define	CRDELAY	030000
#define	VTDELAY	040000
#define DUMC	0100000	/* KLH: Don't Use Modem Control. */

/* Speeds (mostly for DH11) */
#define B0	00	/* 0 baud, for turning lines off */
#define B50	01
#define B75	02
#define B110	03	/* 110 baud */
#define B134	04	/* 134.5 baud */
#define B150	05
#define B200	06
#define B300	07
#define B600	010
#define B1200	011
#define B1800	012
#define B2400	013
#define B4800	014
#define B9600	015
#define BEXTA	016
#define BEXTB	017

/* Hardware bits */
#define	DONE	0200
#define	IENABLE	0100

/* Internal state bits */
#define	TIMEOUT	01		/* Delay timeout in progress */
#define	WOPEN	02		/* Waiting for open to complete */
#define	ISOPEN	04		/* Device is open */
#define	SSTART	010		/* Has special start routine at addr */
#define	CARR_ON	020		/* Software copy of carrier-present */
#define	BUSY	040		/* Output in progress */
#define	ASLEEP	0100		/* Wakeup when output done */
#define ERMODE	0200		/* Erase mode, deleting chars */
#define ACKWAIT	0400		/* timeout is pending */
#define ESCRCV	01000		/* set when escape has been recieved */
#define ENQRCV	02000		/* ENQ recieved by ttyinput */
#define	FROZEN	04000		/* tty output is frozen (by ^F)	*/
#define HELD	010000		/* tty input is held */

#ifdef DNTTY
/* DN-specific TTY flags, in t_dnflgs */
#define	BAUDOT	01		/* When set, TTY is baudot tty. */
#define	FIGS	02		/* Set during FIGS mode, off during LETS */
#define	CTS	04		/* Copy of DM hardware CTS */
#define TMORTS	010		/* Set during RTS-drop timeout */
#define TMOCTS	020		/* Set during CTS-on timeout wait */
#define LICCR	040		/* Last input char was CR */
#define LICLF	0100		/* Last input char was LF */
#endif

/* ttymode parameters */

#define TFLUSH	0020000		/* Request to flush buffer first */
#define TCNTL	0140000		/* Control bits in a ttycommand  */
#define TBYTNUM 0017400		/* Byte number           "       */
#define TBITS	0000377		/* Argument bits         "       */
#define	TCLEAR	0040000		/* Clear control command in a ttycommand */
#define	TSET	0100000		/* Set		   "			 */
#define	TREPLAC	0140000		/* Replace	   "			 */
#define	TGET	0		/* Get		   "			 */

/* t_tflag2 flags */

#define	SILENT	01		/* Silences tty */
#define	LITIN	02		/* 8 bit input */
#define	LITOUT	04		/* 8 bit output */
#define	MORE	010		/* Enables "MORE" processing */
#define EEI	020		/* Explicitly Enable Interrupts
				 * (allows ^Q,^B,^A in raw mode */
#define RACK	040
#define WACK	0100
#define SECRX	0200		/* don't transmit if SECRX is 0 */
#define WHOLD	0400		/* transmit t_hold and t_release */
				/* when the input queue grows/shrinks */

/* Canonical terminal types */

#  define TC_PRINTING   0
#  define TC_VT52       1
#  define TC_ARDS	2
#  define TC_SARDS	3
#  define TC_GSI        4       /* courrier wheel */
#  define TC_GSIE       5       /* elite wheel */
#  define TC_GSIP       6       /* pica wheel */
#  define TC_GSIQ	7
#  define TC_GSIQE	8
#  define TC_GSIQP	9
#  define TC_GT40       10      /* assumed to be running the VT52 scroller */
#  define TC_TERM       11      /* the chris termanal */
#  define TC_CRT        12      /* cathode-ray decwriter */
#  define TC_AJ		13	/* Anderson-Jacobson 832 (Qume)		*/
#  define TC_AJE	14
